crates/integrations/datafusion/src/physical_plan/project.rs

```diff
@@ -0,0 +1,557 @@
+use datafusion::arrow::datatypes::{DataType, Schema as ArrowSchema};
+use datafusion::common::Result as DFResult;
+use datafusion::error::DataFusionError;
+use datafusion::physical_expr::PhysicalExpr;
+use datafusion::physical_expr::expressions::Column;
+use datafusion::physical_plan::projection::ProjectionExec;
+use datafusion::physical_plan::{ColumnarValue, ExecutionPlan};
+use iceberg::arrow::record_batch_projector::RecordBatchProjector;
+use iceberg::arrow::schema_to_arrow_schema;
+use iceberg::spec::{PartitionSpec, Schema};
+use iceberg::table::Table;
+use iceberg::transform::BoxedTransformFunction;
+
+use crate::to_datafusion_error;
+
+/// Column name for the combined partition values struct
+const PARTITION_VALUES_COLUMN: &str = "_partition";
+
+/// Extends an ExecutionPlan with partition value calculations for Iceberg tables.
+///
+/// This function takes an input ExecutionPlan and extends it with an additional column
+/// containing calculated partition values based on the table's partition specification.
+/// For unpartitioned tables, returns the original plan unchanged.
+///
+/// # Arguments
+/// * `input` - The input ExecutionPlan to extend
+/// * `table` - The Iceberg table with partition specification
+///
+/// # Returns
+/// * `Ok(Arc<dyn ExecutionPlan>)` - Extended plan with partition values column
+/// * `Err` - If partition spec is not found or transformation fails
+pub fn project_with_partition(
+    input: Arc<dyn ExecutionPlan>,
+    table: &Table,
+) -> DFResult<Arc<dyn ExecutionPlan>> {
+    let metadata = table.metadata();
+    let partition_spec = metadata.default_partition_spec();
+    let table_schema = metadata.current_schema();
+
+    if partition_spec.is_unpartitioned() {
+        return Ok(input);
+    }
+
+    let input_schema = input.schema();
+
+    // Validate that input schema matches the table schema
+    validate_schema_compatibility(&input_schema, table_schema.as_ref())?;
+
+    let partition_type = build_partition_type(partition_spec, table_schema.as_ref())?;
+    let calculator = PartitionValueCalculator::new(
+        partition_spec.as_ref().clone(),
+        table_schema.as_ref().clone(),
+        partition_type,
+    )?;
+
+    let mut projection_exprs: Vec<(Arc<dyn PhysicalExpr>, String)> =
+        Vec::with_capacity(input_schema.fields().len() + 1);
+
+    for (index, field) in input_schema.fields().iter().enumerate() {
+        let column_expr = Arc::new(Column::new(field.name(), index));
+        projection_exprs.push((column_expr, field.name().clone()));
+    }
+
+    let partition_expr = Arc::new(PartitionExpr::new(calculator));
+    projection_exprs.push((partition_expr, PARTITION_VALUES_COLUMN.to_string()));
+
+    let projection = ProjectionExec::try_new(projection_exprs, input)?;
+    Ok(Arc::new(projection))
+}
+
+/// PhysicalExpr implementation for partition value calculation
+#[derive(Debug, Clone)]
+struct PartitionExpr {
+    calculator: Arc<PartitionValueCalculator>,
```

Comment:
1. Do we still need to `PartitionValueCalculator` struct?
2. Why we need this `Arc`?